pyquil/api/_qpu.py

```diff
@@ -71,35 +72,71 @@ def _extract_bitstrings(ro_sources: List[Optional[Tuple[int, int]]],
 
-        :param endpoint: Address to connect to the QPU server.
+        :param endpoint: Address to connect to the QPU server. If not provided, the
+            endpoint provided by engagement with dispatch is used. One or both must be
+            available and valid.
         :param user: A string identifying who's running jobs.
         :param priority: The priority with which to insert jobs into the QPU queue. Lower
                          integers correspond to higher priority.
+        :param config: PyQuilConfig object, which provides endpoint & engagement values
         """
+        if config:
+            self.config = config
+        else:
+            self.config = PyquilConfig()
 
-        if endpoint is None:
-            raise RuntimeError("""It looks like you've tried to run a program against a QPU but do
- not currently have a reservation on one. To reserve time on Rigetti
- QPUs, use the command line interface, qcs, which comes pre-installed
- in your QMI. From within your QMI, type:
-
-    qcs reserve --lattice <lattice-name>
-
-For more information, please see the docs at
-https://www.rigetti.com/qcs/docs/reservations or reach out to Rigetti
-support at support@rigetti.com.""")
-
-        self.client = Client(endpoint)
+        self.endpoint = endpoint
         self.user = user
         self._last_results: Dict[str, np.ndarray] = {}
         self.priority = priority
 
         super().__init__()
 
+    def build_client(self) -> Client:
+        endpoint = self.endpoint or self.config.qpu_url
+        if endpoint is None:
+            raise UserMessageError(
+                """It looks like you've tried to run a program against a QPU but do
+                not currently have a reservation on one. To reserve time on Rigetti
+                QPUs, use the command line interface, qcs, which comes pre-installed
+                in your QMI. From within your QMI, type:
+
+                    qcs reserve --lattice <lattice-name>
+
+                For more information, please see the docs at
+                https://www.rigetti.com/qcs/docs/reservations or reach out to Rigetti
+                support at support@rigetti.com.""")
+
+        logger.debug(f"QPU Client connecting to {endpoint}")
+
+        return Client(endpoint, auth_config=self.client_auth_config)
+
+    @property
+    def client(self) -> Client:
+        if not (self.engagement and self.engagement.is_valid() and self._client):
+            self._client = self.build_client()
+        return self._client
+
+    @property
+    def client_auth_config(self) -> Optional[ClientAuthConfig]:
+        if self.engagement is not None:
+            return ClientAuthConfig(
+                client_public_key=self.engagement.client_public_key,
+                client_secret_key=self.engagement.client_secret_key,
+                server_public_key=self.engagement.server_public_key)
+
+    @property
+    def engagement(self) -> Engagement:
+        return self.config.engagement
```

Comment:
maybe make these private? will users access them?

Reply:
It doesn't seem like a common use case, but I can imagine situations in which a user would want to provide this information manually (i.e. cache & reload it between script runs, or sharing an engagement between collaborating users)

Reply:
I don't currently see a use case for this property -- in general, i don't think we should use properties unless we have to (for backwards compatibility). looks like you can just use `self.config.engagement` everywhere